src/main/java/ch/njol/skript/expressions/ExprBookPages.java

```diff
@@ -19,99 +19,189 @@
-@Examples({"on book sign:",
-			"\tmessage \"Book Pages: %pages of event-item%\"",
-			"\tmessage \"Book Page 1: %page 1 of event-item%\""})
-@Since("2.2-dev31")
+@Examples({
+	"on book sign:",
+	"\tmessage \"Book Pages: %pages of event-item%\"",
+	"\tmessage \"Book Page 1: %page 1 of event-item%\"",
+	"set page 1 of player's held item to \"Book writing\""
+})
+@Since("2.2-dev31, INSERT VERSION (changers)")
 public class ExprBookPages extends SimpleExpression<String> {
-	
+
 	static {
 		Skript.registerExpression(ExprBookPages.class, String.class, ExpressionType.PROPERTY,
-				"[all] [the] [book] (pages|content) of %itemtypes%",
-				"%itemtypes%'s [book] (pages|content)",
-				"[book] page %number% of %itemtypes%", 
-				"%itemtypes%'s [book] page %number%");
+			"[(all [[of] the]|the)] [book] (pages|content) of %itemtypes%",
+			"%itemtypes%'[s] [book] (pages|content)",
+			"[book] page %number% of %itemtypes%",
+			"%itemtypes%'[s] [book] page %number%");
 	}
-	
-	private static final ItemType bookItem = Aliases.javaItemType("book with text");
-	
-	@SuppressWarnings("null")
-	private Expression<ItemType> book;
+
+	private Expression<ItemType> items;
 	@Nullable
 	private Expression<Number> page;
-	
-	@SuppressWarnings("null")
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		if (matchedPattern == 0 || matchedPattern == 1) {
+			items = (Expression<ItemType>) exprs[0];
+		} else {
+			if (matchedPattern == 2) {
+				page = (Expression<Number>) exprs[0];
+				items = (Expression<ItemType>) exprs[1];
+			} else {
+				items = (Expression<ItemType>) exprs[0];
+				page = (Expression<Number>) exprs[1];
+			}
+		}
+		return true;
+	}
+
+	@Override
 	@Nullable
+	@SuppressWarnings("null")
+	protected String[] get(Event event) {
+		List<String> pages = new ArrayList<>();
+		for (ItemType itemType : items.getArray(event)) {
+			if (itemType == null || !(itemType.getItemMeta() instanceof BookMeta))
+				continue;
+			BookMeta bookMeta = (BookMeta) itemType.getItemMeta();
+			if (isAllPages()) {
+				pages.addAll(bookMeta.getPages());
+			} else {
+				Number pageNumber = page.getSingle(event);
+				if (pageNumber == null)
+					continue;
+				int page = pageNumber.intValue();
+				if (page <= 0 || page > bookMeta.getPageCount())
+					continue;
+				pages.add(bookMeta.getPage(page));
+			}
+		}
+		return pages.toArray(new String[0]);
+	}
+
 	@Override
-	protected String[] get(Event e) {
-		ItemStack itemStack = book.getSingle(e).getRandom();
-		if (itemStack == null || !bookItem.isOfType(itemStack))
-			return null;
-		List<String> pages = ((BookMeta) itemStack.getItemMeta()).getPages();
-		if (page != null){
-			Number pageNumber = page.getSingle(e);
-			if (pageNumber == null){
+	@Nullable
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		switch (mode) {
+			case RESET:
+			case DELETE:
+				return CollectionUtils.array();
```

Comment:
```suggestion
			case RESET:
			case DELETE:
				return CollectionUtils.array();
```
null can just be returned

Reply:
Returning null there will make it so you can't use those changers

Reply:
sorry I confused myself w/ the other method